Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -69,46 +71,192 @@ class CDMPrivateOpenCDM : public CDMPrivate {
+    }
+
+    static void didGenerateChallenge(OpenCDMSession*, const char[], const uint8_t[], const uint16_t);
+    static void didUpdateKey(OpenCDMSession*, const uint8_t[], const uint8_t);
+    static void didReceiveMessage(OpenCDMSession*, const char[]);
+    static bool tryNotify(OpenCDMSession*, Notification, const uint8_t[], uint16_t);
+    static void notify(OpenCDMSession*, Notification, const uint8_t[], uint16_t);
+
+    friend class NeverDestroyed<OCDMSessionCallbacksNotifier>;
+    OpenCDMSessionCallbacks m_calbacks;
+    HashMap<OpenCDMSession*, Observer*> m_observers;
+    GenericTaskQueue<Timer> m_taskQueue;
+};
+
+OCDMSessionCallbacksNotifier& OCDMSessionCallbacksNotifier::singleton()
+{
+    static NeverDestroyed<OCDMSessionCallbacksNotifier> instance;
+    return instance;
+}
+
+void OCDMSessionCallbacksNotifier::addObserver(OpenCDMSession* session, Observer* observer)
+{
+    m_observers.set(session, observer);
+}
+
+void OCDMSessionCallbacksNotifier::removeObserver(Observer* observer)
+{
+    m_observers.removeIf([observer](const auto& keyAndValue) { return observer == keyAndValue.value; });
+}
 
-    bool isValid() const { return m_isValid; }
+bool OCDMSessionCallbacksNotifier::tryNotify(OpenCDMSession* session, Notification method, const uint8_t message[], uint16_t messageLength)
+{
+    if (!isMainThread()) {
+        // Make sure all happens on the main thread to avoid locking.
+        callOnMainThread([session, method, message, messageLength]() {
+            OCDMSessionCallbacksNotifier::tryNotify(session, method, message, messageLength);
+        });
+        return true;
+    }
+
+    OCDMSessionCallbacksNotifier& self = OCDMSessionCallbacksNotifier::singleton();
+    auto entry = self.m_observers.find(session);
+    if (entry != self.m_observers.end()) {
+        (entry->value->*method)(std::string(&message[0], &message[messageLength]));
+        return true;
+    }
+
+    return false;
+}
+
+void OCDMSessionCallbacksNotifier::notify(OpenCDMSession* session, Notification method, const uint8_t message[], uint16_t messageLength)
+{
+    if (OCDMSessionCallbacksNotifier::tryNotify(session, method, message, messageLength))
+        return;
+
+    // It looks like there's no observer interested in the notification. It might be due to the fact it hasn't had chance
+    // to register yet because e.g. the notification is sent synchronously.
+    // Currently this happens for opencdm_create_session() which synchronously calls the process_challenge() callback and
+    // the only way to figure out who is the receiver is by OpenCMDSession. Problem however is that the caller can't know
+    // what's its sesssion at this point - it's creating one now.
+    auto& self = OCDMSessionCallbacksNotifier::singleton();
+    self.m_taskQueue.enqueueTask([session, method, message, messageLength]() {
+        OCDMSessionCallbacksNotifier::tryNotify(session, method, message, messageLength);
+    });
```

Comment:
I'm not confident I understand what's going on here. Isn't it a bug that process_challenge() is called synchronously in this case? Seems like depending on the timer resolution, we could still miss messages here during session setup. I don't know what to suggest about fixing it.

Reply:
Yeah. It's something to look into at OCDM side. What currently happens is:
Session upon creation calls ocdm's create_session() which returns OpenCDMSession* but synchronously calls process_challenge() passing OpenCDMSession* to it. Session isn't yet registered to the notifications because it has to have OpenCDMSession* for that so it registers after create_session() returns so after the notification. This is why the Notifier - when it detects there is no observer - calls the notification again asynchonously so that the second call arrives when the Session is already registered. I believe that nothing will be missed due to the fact all happens on the main thread so Session ctor has to finish for the asynchronous notification to run meaning it'll happen after registration.

Reply:
OK, we'll see :) I am happy to leave a FIXME to revisit this at some point and get your work landed.

Reply:
FIXME added